### ISSUE TITLE
[changelog] Missing and duplicate entries

### ISSUE THREAD

**Issue by alloy** [react-native-community/releases#160]
Related to [this card](https://github.com/react-native-community/discussions-and-proposals/projects/1#card-31964640).

There are currently a few issues with the changelog generator:

* No more than 250 commits are considered due to GitHub’s [“compare commits” API](https://developer.github.com/v3/repos/commits/#compare-two-commits) being used. This should be replaced by usage of an API that is able to iterate over long commit ranges and require an authentication token, for the user running the script, to allow paginating through all the commits.

* Because of how RN’s release process works, we cannot simply use git’s history to inform the generator on what changes should be included in the changelog.

  The reason being that cherry-picked commits are new changes, as far as git is concerned, and may thus be picked up again in subsequent releases. (For a reproduction of this issue, see [this example](https://github.com/alloy/RNBranchingStrategyRepro/compare/v1...v2). The `v1` commit is not expected to show up in this diff, as it was picked for `v1`, yet there it is.)

  After having slept on it for a few days, I think the best way to solve this is to:
  1. Use the “differential revision” that FB’s infrastructure adds to each commit to find the SHA of the commit in `master`, then;
  1. check if that SHA already exists in the changelog, in which case it is skipped, or otherwise;
  1. add the entry to the changelog with the commit SHA as it exists in `master`, rather than the stable branch.
  1. Finally, when invoking the generator, compare the current `0.x-stable` branch against the commit from which the _previous_ `0.x-stable` branch was forked off of `master`–as not all commits in `master` since then will have been included in the previous release.

TL;DR Use the `master` commit ref as canonical and check for its existence in the changelog to decide wether or not to include it.

**Comment by TheSavior** [comment]
Sounds like good approaches to me! Thanks for working on this!
